### ISSUE TITLE
Pipeline is not concurrency-safe with streams

### ISSUE THREAD

**Issue by kak-tus** [go-redis/redis#990]
I use Cluster client with pipelining.
Then start 4 goroutines, each goroutine get pipeline, and do 100000 xadd command in pipeline.
After that I check stream length and got not exactly 400000 messages in it, but 402384 (or near this).

How to reproduce.

Start cluster
```
docker run --rm -it --network host redis:alpine --cluster-enabled yes --cluster-config-file nodes.conf --port 7001
docker run --rm -it --network host redis:alpine --cluster-enabled yes --cluster-config-file nodes.conf --port 7002
docker run --rm -it --network host redis:alpine --cluster-enabled yes --cluster-config-file nodes.conf --port 7003
docker run --rm -it --network host redis:alpine --cluster-enabled yes --cluster-config-file nodes.conf --port 7004
docker run --rm -it --network host redis:alpine --cluster-enabled yes --cluster-config-file nodes.conf --port 7005
docker run --rm -it --network host redis:alpine --cluster-enabled yes --cluster-config-file nodes.conf --port 7006
```

Init cluster
```
docker run --rm -it --network host redis:alpine redis-cli --cluster create 172.17.0.1:7001 172.17.0.1:7002 172.17.0.1:7003 172.17.0.1:7004 172.17.0.1:7005 172.17.0.1:7006 --cluster-replicas 1
```

Run script
```
package main

import (
	"sync"

	"github.com/go-redis/redis"
)

func main() {
	rd := redis.NewClusterClient(&redis.ClusterOptions{
		Addrs:    []string{"172.17.0.1:7001", "172.17.0.1:7002"},
	})

	wg := sync.WaitGroup{}

	for j := 0; j < 4; j++ {
		wg.Add(1)
		go func() {
			pipe := rd.Pipeline()

			for i := 0; i < 100000; i++ {
				pipe.XAdd(&redis.XAddArgs{
					Stream: "test-stream",
					ID:     "*",
					Values: map[string]interface{}{"m": "{}"},
				})
			}

			_, err := pipe.Exec()
			if err != nil {
				println(err.Error())
			}

			wg.Done()
		}()
	}

	wg.Wait()
}
```

Check stream length
```
docker run --rm -it --network host redis:alpine redis-cli -c -h 172.17.0.1 -p 7003 xinfo stream test-stream
```

If you got correct stream length - please try to set greater concurrency.

**Comment by kak-tus** [comment]
Same with lpush. Pushed 400000, llen 403971.

**Comment by kak-tus** [comment]
Hm. May be problem is not with concurrency, but with pipelining.

1. If using separated redis client in each goroutine - same problem.
2. If using single command mode (change pipe.XAdd (or pipe.LPush) to rd.XAdd (or rd.LPush)) and separated redis clients for each goroutine - all ok.
3. If using single command mode and shared redis client - all ok.

**Comment by kak-tus** [comment]
I've found the problem, but it is really unsuddenly.

In case of pipeline with many commands - it produces a long time communications with Redis.
Default write and read timeout in go-redis is not so big.
But usual Pipeline works in nontransactional way.
And in case ot write or read timeout - go-redis do silent retransmission.

This problem can be fixed with:
1. User must set bigger Read/Write timeout if he want's to produce such big pipelines. In my case I set 60 seconds read/write timeout (for remote Redis).
2. To ignore retransmissions user must use TxPipeline, so in case of retransmission - lost transactions will be ignored.

I don't close issue, because I think, it is good idea to describe this in Pipeline and TxPipeline docs.
